### PR TITLE
Implement P2467R1 `ios_base::noreplace`: Exclusive Mode For `fstream`s (#2932)

### LINKED ISSUE #2932: P2467R1 `ios_base::noreplace`: Exclusive Mode For `fstream`s
[P2467R1](https://wg21.link/P2467R1) `ios_base::noreplace`: Exclusive Mode For `fstream`s

Feature-test macro (expected):
```cpp
#define __cpp_lib_ios_noreplace 202207L
```

### PR DESCRIPTION
**Pull request by Atari2**
As was pointed out by @StephanTLavavej, since this PR has changes to `fiopen.cpp`, it will be `affects redist`. 
While writing the tests I tried to avoid having to change things that could affect binary compatibility but it seemed like the only option was to add the now-valid entries to the array in `_Xfiopen`, otherwise the calls to it would just fail.

Fixes #2932.

### CODE REVIEW

File: stl/src/fiopen.cpp
Comment (MattStephanson) [reply stl/src/fiopen.cpp:49]: Browsing through the UCRT source, it looks like this is a documentation issue. `_fsopen` is a wrapper around `_openfile`, which will translate `'x'` to `_O_EXCL`. This eventually calls down to `_wsopen_nolock`, which translates `_O_EXCL` to `OPEN_EXISTING` for `CreateFileW`. If this is right, I think it would be better to use that flag, not try to check for file existance ourselves, and avoid the TOCTOU issue.

**Comment by StephanTLavavej** [comment]
Thanks! I updated your PR description to say "Fixes \<issue number\>", which tells GitHub to link the tracking issue and close it when this PR is merged.

File: stl/src/fiopen.cpp
Comment (Atari2) [reply stl/src/fiopen.cpp:50]: I pushed changes, should be resolved now, hopefully, if anything else is needed/discovered, please feel free to ping me on discord.

**Review by strega-nil-ms — APPROVED** [review]
LGTM! Thanks!

File: tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp:10]: In modern C++ style, I strongly prefer to use `string_view`s to macros for string constants. That said, this is only ever used to initialize `path`s, so this could as well be a `path` constant, simplifying the rest of the code a bit.

File: tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp:15]: Meganit: there's no need to name this object, it could simply be a temporary.

File: tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp:23]: Also meganit (I don't feel bad since I've committed myself to applying these cleanups instead of making you do it) the parameters to these functions could be `const` to ensure we don't accidentally modify them locally.

**Review by CaseyCarter — APPROVED** [review]
Just style nits with the test, for which I'll push a cleanup.

File: tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp:66]: There's one more failure scenario we can test: `IB::in | IB::binary | IB::noreplace`. This will complete the matrix of "`in | noreplace` with all possible settings of `binary` and `trunc`".

For clarity, I would also like to reorder `in` to be first (following the style of `out` above), so that the pattern is clearer.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks @Atari2, this looks great! I've pushed a merge with `main` in order to resolve trivial adjacent-add conflicts with the Modules PR that just landed (in `yvals_core.h` and `tests/std/test.lst`, as the paper numbers and macro names sorted close together), followed by fine-grained commits to address minor things I noticed. (The most significant being a pre-emptive fix for the `/clr:pure` internal build.) Also FYI @CaseyCarter @strega-nil-ms as I pushed after you approved.

While some `affects redist` PRs can be merged for locked-redist releases (when the redist's behavior changes, but we're comfortable with the compatibility implications), this feature *requires* an updated redist or it'll fail at runtime, so we'll place this PR in cryo-stasis (`blocked` status) until changes start flowing into VS 2022 17.6. I believe that this will happen at the end of October (note: this has no relationship to public release dates of Previews - it is purely when internal merges switch over from one release branch to another).

**Comment by StephanTLavavej** [comment]
VS 2022 17.6 Preview 1's redist will be unlocked, so we should be able to mirror this now. I've pushed a conflict-free merge with `main` to pick up the ~2 months of changes there (and submodule updates), to be careful.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this feature in a binary-compatible way! :heart_eyes_cat: :rocket: :tada:

**Pull request merged**
